ap/inference/__init__.py

```diff
@@ -0,0 +1,1405 @@
+            self.init_func()
+
+    def to_yaml(self, stream: Optional[TextIO] = None) -> Union[None, str]:
+        """
+        Writes the content of the :py:attr:`model` into a file-like object *stream* when given, and
+        returns a string representation otherwise.
+        """
+        return yaml.dump(self.model, stream=stream, Dumper=self.YamlDumper)
+
+    def pprint(self) -> None:
+        """
+        Pretty-prints the content of the :py:attr:`model` in yaml-style.
+        """
+        print(self.to_yaml())
+
+    #
+    # property access to top-level objects
+    #
+
+    @property
+    def categories(self) -> DotDict:
+        return self.model.categories
+
+    @property
+    def parameter_groups(self) -> DotDict:
+        return self.model.parameter_groups
+
+    #
+    # handling of categories
+    #
+
+    def get_categories(
+        self,
+        category: Optional[Union[str, Sequence[str]]] = None,
+        only_names: bool = False,
+    ) -> List[Union[DotDict, str]]:
+        """
+        Returns a list of categories whose name match *category*. When *only_names* is *True*, only
+        names of categories are returned rather than structured dictionaries.
+        """
+        # rename arguments to make their meaning explicit
+        category_pattern = category
+
+        match = pattern_matcher(category_pattern or "*")
+        return [
+            (category.name if only_names else category)
+            for category in self.categories
+            if match(category.name)
+        ]
+
+    def get_category(
+        self,
+        category: Union[str, Sequence[str]],
+        only_name: bool = False,
+        silent: bool = False,
+    ) -> Union[DotDict, str]:
+        """
+        Returns a single category whose name matches *category*. An exception is raised if no or
+        more than one category is found, unless *silent* is *True* in which case *None* is returned.
+        When *only_name* is *True*, only the name of the category is returned rather than a
+        structured dictionary.
+        """
+        # rename arguments to make their meaning explicit
+        category_name = category
+
+        categories = self.get_categories(category_name, only_names=only_name)
+
+        # length checks
+        if not categories:
+            if silent:
+                return None
+            raise ValueError(f"no category named '{category_name}' found")
+        if len(categories) > 1:
+            if silent:
+                return None
+            raise ValueError(f"category name '{category_name}' matches more than one category")
+
+        return categories[0]
+
+    def has_category(
+        self,
+        category: Union[str, Sequence[str]],
+    ) -> bool:
+        """
+        Returns *True* if a category whose name matches *category* is existing, and *False*
+        otherwise.
+        """
+        # rename arguments to make their meaning explicit
+        category_pattern = category
```

Comment:
I actually used that before, but when writing the test model, it felt oddly cumbersome to always add `*_pattern` to all keyword arguments. In internally in the implementation, I found it easier to rename the variables first so that one knows what type of string one is dealing with.

Reply:
Since even e.g. `category_pattern` would be incorrect since also lists or sequences of patterns are accepted, I think it might be good to explicitly write this in the doc strings as you suggested.

Reply:
I added these hints to all doc strings 👍